networkx/algorithms/approximation/density.py

```diff
@@ -0,0 +1,166 @@
+        The algorithm to use to approximate the densest subgraph.
+        Supported options: 'greedy++'.
+        Must be specified as a keyword argument. Other inputs produce a ValueError.
+
+    Returns
+    -------
+    d : float
+        The density of the approximate subgraph found.
+
+    S : set
+        The subset of nodes defining the approximate densest subgraph.
+
+    Examples
+    --------
+    >>> G = nx.star_graph(4)
+    >>> nx.approximation.densest_subgraph(G, iterations=1)
+    (0.8, {0, 1, 2, 3, 4})
+
+    Notes
+    -----
+    For a subset of the nodes of $G$, $S \subseteq V(G)$, define
+    $E(S) = \{ (u,v) : (u,v)\in E(G), u\in S, v\in S \}$ as the set of
+    edges with both endpoints in $S$. The density of $S$ is defined as
+    $E(S)/|S|$, the ratio between the edges in the subgraph $G[S]$ and
+    the number of nodes in that subgraph. The densest subgraph problem
+    asks to find the subgraph $S \subseteq V(G)$ with maximum density.
+
+    The densest subgraph problem is polynomial time solvable using
+    maximum flow, commonly refered to as Goldberg's algorithm.
+    However, the algorithm is quite involved. It first
+    binary searches on the optimal density, $d^\ast$. For a guess of
+    the density $d$, it sets up a flow network $G'$ with size O(m). The
+    maximum flow solution either informs the algorithm that no subgraph
+    with density $d$ exists, or it returns (implicitly) a subgraph with density
+    $d$. However, this is inherently bottlenecked by the maximum flow algorithm.
+    For example, authors from [2]_ noted that Goldberg’s algorithm
+    failed on many large graphs even though they used a highly optimized
+    maximum flow library.
+
+    Due to the importance of the problem and solving it on large scale graphs,
+    there are several known approximation algorithms for the problem.
+
+    Charikar [1]_ described a very simple 1/2-approximation algorithm for DSG known as
+    the greedy "peeling" algorithm. The algorithm creates an ordering of the nodes as follows.
+    The first node $v_1$ is the one with the smallest degree in $G$ (ties broken arbitrarily).
+    It selects $v_2$ to be the smallest degree node in $G \setminus v_1$. Letting $G_i$ be the
+    graph after removing $v_1, ..., v_i$ (with $G_0=G$), the algorithm returns the graph among
+    $G_0, ..., G_n$ with the highest density.
+
+    Boob et al. [2]_ generalized this algorithm into Greedy++, an iterative algorithm that runs
+    several rounds of "peeling". In fact, Greedy++ with 1 iteration is precisely Charikar's algorithm.
+    The algorithm converges to a $(1-\epsilon)$ approximate densest
+    subgraph in $O(\Delta(G)\log n/\epsilon^2)$ iterations, and has other desirable properties as shown by
+    [4]_ and [5]_.
+
+    Harb et al. [3]_ gave a faster and more scalable algorithm using ideas from quadratic programming
+    for the densest subgraph, which is based on FISTA algorithm.
```

Comment:
maybe spell out FISTA?
```suggestion
    for the densest subgraph, which is based on a fast iterative
    shrinkage-thresholding algorithm (FISTA) algorithm.
```